soc: stm32n6: Add userspace support

erwango opened the pull request:
Add userspace support on stm32n6 SoC.

Involves:
- As based on Cortex-M55, implies enabling MPU_GAP_FILLING on the SoC series
- In XIP mode, this requires being able to compute FLASH_SIZE from external flash properties
- In !XIP mode (which is the mode used in CI), this implies declaring RAM_RW and RAM_RO static regions, ~~which requires the ability to provide static regions in run time. This possibility is proposed as a new `MPU_STATIC_REGIONS_DYNAMIC_INIT` option.~~ done with a dedicated linker snippet.

Note: Requires https://github.com/zephyrproject-rtos/zephyr/pull/95152 to build in XIP

erwango committed:
soc: stm32: stm32n6: Define MPU_GAP_FILLING also with USERSPACE

MCU_GAP_FILLING is advised on ARMv8 based platforms to prevent from
against attacks that attempt to execute malicious code from SRAM.
It is enabled by default when USERSPACE is disabled, enable it
also when enabled.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
soc: stm32n6: Compute FLASH_SIZE when XIP=y

In XIP, we need to know FLASH_SIZE.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
include: arch: arm: cortex_m: Provide __rom_region_size linker symbol

__rom_region_size linker symbol is available for other archs, but was
missing for Cortex-M. Add it now.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

wearyzen reviewed (changes requested):
> In some cases, static mpu regions definitions require computation on
> symbols that are not possible  at build time and then can't be provided
> in usual mpu_config struct.

If by some cases you mean addresses generated by linker scripts then this is not the right approach. Please have a look at other soc mpu_configs that are already using the same e.g. https://github.com/zephyrproject-rtos/zephyr/blob/67f246477f93ac8e7bd4b953a9f47d4f5f5b4ef5/soc/arm/fvp_aemv8r/aarch32/arm_mpu_regions.c#L31

erwango commented:
> Ah ok, let me see if there is any existing way to avoid this. There are already too many places where mpu_config takes place so I'm trying to avoid yet another new way to do it.

Ok, thanks. This is the only way I've found out to make it work, but I'm open to alternate suggestions.

> BTW, just for my knowledge, why can't the regions be reserved at compile time with either DT or KCONFIGs?

!XIP is a common configuration on STM32N6 (and this configuration is also specifically used for our tests in CI).
To make !XIP work with USERSPACE, I need to provide configuration for RAM_RO and RAM_RW regions, which implies to make use of linker symbols (`__rom_region_start`, `__rom_region_size`, `_image_ram_start`).

erwango committed:
soc: st: stm32n6: Provide static mpu regions for !XIP

In !XIP case, provide RAM_RO and RAM_RW regions with their
respective MPU configurations.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>
Co-authored-by: Sudan Landge <sudan.landge@arm.com>

etienne-lms commented on soc/st/stm32/stm32n6x/Kconfig:
```diff
@@ -20,6 +20,8 @@ config SOC_SERIES_STM32N6X
 	select SOC_EARLY_INIT_HOOK
 	select TRUSTED_EXECUTION_SECURE
 	select BUILD_OUTPUT_BIN
+	# MPU_GAP_FILLING is default when !USERSPACE, select it in the other case as well.
+	select MPU_GAP_FILLING if USERSPACE
```
Not done, but maybe that's on purpose.
